File: tests/python/test_tokenizer.py
"""Byte-exact parity tests for esm.cpp's Tokenizer against HF EsmTokenizer.

Phase 0, Slice 2 acceptance criterion: 10,000 random sequences tokenize
byte-exact to HuggingFace `facebook/esm2_t6_8M_UR50D`.
"""

from __future__ import annotations

import random

import pytest

import esm_cpp

# HF is optional at test time; skip the parity test if it isn't installed.
try:
    from transformers import AutoTokenizer  # type: ignore[import-not-found]

    HF_AVAILABLE = True
except ImportError:  # pragma: no cover
    HF_AVAILABLE = False


CANONICAL_AA = "LAGVSERTIDPKQNFYMHWC"  # 20 amino acids, UR50 order
RARE_AA = "XBUZO.-"
NOISE_CHARS = "0123456789abcdefghijklmnopqrstuvwxyz!@#$%^&*"


@pytest.fixture(scope="module")
def hf_tokenizer():
    if not HF_AVAILABLE:
        pytest.skip("transformers not installed")
    return AutoTokenizer.from_pretrained("facebook/esm2_t6_8M_UR50D")


@pytest.fixture
def ours():
    return esm_cpp.Tokenizer()


def _random_sequence(rng: random.Random, alphabet: str, lo: int, hi: int) -> str:
    length = rng.randint(lo, hi)
    return "".join(rng.choices(alphabet, k=length))


def test_vocab_constants(ours):
    assert esm_cpp.Tokenizer.vocab_size == 33
    assert esm_cpp.Tokenizer.cls_id == 0
    assert esm_cpp.Tokenizer.pad_id == 1
    assert esm_cpp.Tokenizer.eos_id == 2
    assert esm_cpp.Tokenizer.unk_id == 3
    assert esm_cpp.Tokenizer.mask_id == 32
    assert ours.id_to_token(20) == "M"
    assert ours.token_to_id("M") == 20


@pytest.mark.parametrize("seq,expected", [
    ("MKTGV", [0, 20, 15, 11, 6, 7, 2]),
    ("XBUZ", [0, 24, 25, 26, 27, 2]),
    ("MKTGVBUZO.X-", [0, 20, 15, 11, 6, 7, 25, 26, 27, 28, 29, 24, 30, 2]),
    ("M K T G V", [0, 20, 15, 11, 6, 7, 2]),
    ("mktgv", [0, 3, 2]),
    ("<mask>", [0, 32, 2]),
    ("M<mask>K", [0, 20, 32, 15, 2]),
    ("<null_1>", [0, 31, 2]),
])
def test_known_cases(ours, seq, expected):
    assert ours.encode(seq) == expected


def test_decode_round_trip(ours):
    seq = "MKTGVAQERSILDPQNFYMHWC"
    ids = ours.encode(seq, add_special=False)
    assert ours.decode(ids) == seq


def test_truncation_keeps_special_tokens(ours):
    seq = "M" * 2000
    ids = ours.encode(seq, add_special=True, truncate=True)
    assert len(ids) == esm_cpp.Tokenizer.model_max_length
    assert ids[0] == 0
    assert ids[-1] == 2


@pytest.mark.parametrize("num_seqs", [10_000])
def test_byte_exact_parity_against_hf(ours, hf_tokenizer, num_seqs):
    """The slice-level acceptance criterion."""
    rng = random.Random(0)
    alphabets = [
        CANONICAL_AA,                              # majority: canonical only
        CANONICAL_AA + RARE_AA,                    # canonical + rare
        CANONICAL_AA + RARE_AA + NOISE_CHARS,      # everything: stress unk/whitespace
    ]
    weights = [0.6, 0.3, 0.1]
    mismatches: list[tuple[str, list[int], list[int]]] = []
    for _ in range(num_seqs):
        alphabet = rng.choices(alphabets, weights=weights, k=1)[0]
        seq = _random_sequence(rng, alphabet, 1, 200)
        # truncate=False to match HF's default (no truncation).
        ours_ids = ours.encode(seq, add_special=True, truncate=False)
        hf_ids = hf_tokenizer(seq, add_special_tokens=True)["input_ids"]
        if ours_ids != hf_ids:
            mismatches.append((seq, ours_ids, hf_ids))
            if len(mismatches) > 5:
                break
    if mismatches:
        details = "\n".join(
            f"  seq={s!r}\n    ours={o}\n    hf  ={h}" for s, o, h in mismatches[:5]
        )
        pytest.fail(f"{len(mismatches)} mismatches in {num_seqs} sequences:\n{details}")

File: tests/cpp/test_tokenizer.cpp
#include <gtest/gtest.h>

#include <string>
#include <vector>

#include "esm_cpp/tokenizer.h"

using esm::Tokenizer;

namespace {

std::vector<int32_t> Encode(const Tokenizer& t, std::string_view s,
                            bool add_special = true) {
  return t.Encode(s, add_special, /*truncate=*/true);
}

}  // namespace

TEST(Tokenizer, VocabIsUR50FrequencyOrder) {
  Tokenizer t;
  EXPECT_EQ(t.IdToToken(0), "<cls>");
  EXPECT_EQ(t.IdToToken(1), "<pad>");
  EXPECT_EQ(t.IdToToken(2), "<eos>");
  EXPECT_EQ(t.IdToToken(3), "<unk>");
  // First few canonical amino acids — order matters and is NOT alphabetical.
  EXPECT_EQ(t.IdToToken(4), "L");
  EXPECT_EQ(t.IdToToken(5), "A");
  EXPECT_EQ(t.IdToToken(6), "G");
  EXPECT_EQ(t.IdToToken(7), "V");
  // Spot-check rare aa and gap/null/mask near the end.
  EXPECT_EQ(t.IdToToken(24), "X");
  EXPECT_EQ(t.IdToToken(28), "O");
  EXPECT_EQ(t.IdToToken(29), ".");
  EXPECT_EQ(t.IdToToken(30), "-");
  EXPECT_EQ(t.IdToToken(31), "<null_1>");
  EXPECT_EQ(t.IdToToken(32), "<mask>");
}

TEST(Tokenizer, CanonicalSequenceMatchesHfReference) {
  Tokenizer t;
  // HF: tok("MKTGV", add_special_tokens=True) -> [0, 20, 15, 11, 6, 7, 2]
  EXPECT_EQ(Encode(t, "MKTGV"), (std::vector<int32_t>{0, 20, 15, 11, 6, 7, 2}));
}

TEST(Tokenizer, NoSpecialTokens) {
  Tokenizer t;
  EXPECT_EQ(t.Encode("MKTGV", /*add_special=*/false, /*truncate=*/true),
            (std::vector<int32_t>{20, 15, 11, 6, 7}));
}

TEST(Tokenizer, RareAminoAcids) {
  Tokenizer t;
  // HF: tok("XBUZ", add_special) -> [0, 24, 25, 26, 27, 2]
  EXPECT_EQ(Encode(t, "XBUZ"), (std::vector<int32_t>{0, 24, 25, 26, 27, 2}));
  // HF: tok("MKTGVBUZO.X-", add_special) ->
  //   [0, 20, 15, 11, 6, 7, 25, 26, 27, 28, 29, 24, 30, 2]
  EXPECT_EQ(Encode(t, "MKTGVBUZO.X-"),
            (std::vector<int32_t>{0, 20, 15, 11, 6, 7, 25, 26, 27, 28, 29, 24,
                                  30, 2}));
}

TEST(Tokenizer, WhitespaceAndUnknowns) {
  Tokenizer t;
  // HF: tok("M K T G V") == tok("MKTGV")
  EXPECT_EQ(Encode(t, "M K T G V"),
            (std::vector<int32_t>{0, 20, 15, 11, 6, 7, 2}));
  // HF: tok("mktgv") -> [0, 3, 2] (single <unk>, NOT five)
  EXPECT_EQ(Encode(t, "mktgv"), (std::vector<int32_t>{0, 3, 2}));
  // HF: tok("M0X0K") -> [0, 20, 3, 24, 3, 15, 2]
  EXPECT_EQ(Encode(t, "M0X0K"),
            (std::vector<int32_t>{0, 20, 3, 24, 3, 15, 2}));
  // HF: tok("M00K") -> [0, 20, 3, 15, 2] (consecutive non-vocab collapse)
  EXPECT_EQ(Encode(t, "M00K"), (std::vector<int32_t>{0, 20, 3, 15, 2}));
  // HF: tok("M  K") -> [0, 20, 15, 2] (whitespace stripped, no <unk>)
  EXPECT_EQ(Encode(t, "M  K"), (std::vector<int32_t>{0, 20, 15, 2}));
  // HF: tok("MmK") -> [0, 20, 3, 15, 2]
  EXPECT_EQ(Encode(t, "MmK"), (std::vector<int32_t>{0, 20, 3, 15, 2}));
}

TEST(Tokenizer, MultiCharSpecials) {
  Tokenizer t;
  // HF: tok("<mask>") -> [0, 32, 2]
  EXPECT_EQ(Encode(t, "<mask>"), (std::vector<int32_t>{0, 32, 2}));
  // HF: tok("M<mask>K") -> [0, 20, 32, 15, 2]
  EXPECT_EQ(Encode(t, "M<mask>K"),
            (std::vector<int32_t>{0, 20, 32, 15, 2}));
  // HF: tok("<null_1>") -> [0, 31, 2]
  EXPECT_EQ(Encode(t, "<null_1>"), (std::vector<int32_t>{0, 31, 2}));
  // HF: tok("M<>K") -> [0, 20, 3, 15, 2] (partial '<' is <unk>)
  EXPECT_EQ(Encode(t, "M<>K"), (std::vector<int32_t>{0, 20, 3, 15, 2}));
}

TEST(Tokenizer, EmptyAndAllWhitespace) {
  Tokenizer t;
  EXPECT_EQ(Encode(t, ""), (std::vector<int32_t>{0, 2}));
  EXPECT_EQ(Encode(t, "   "), (std::vector<int32_t>{0, 2}));
  EXPECT_EQ(Encode(t, "\t\n"), (std::vector<int32_t>{0, 2}));
}

TEST(Tokenizer, TruncationKeepsClsAndEos) {
  Tokenizer t;
  std::string long_seq(2000, 'M');
  auto ids = t.Encode(long_seq, /*add_special=*/true, /*truncate=*/true);
  ASSERT_EQ(ids.size(), Tokenizer::kModelMaxLength);
  EXPECT_EQ(ids.front(), Tokenizer::kClsId);
  EXPECT_EQ(ids.back(), Tokenizer::kEosId);
  // Interior tokens are all M (id 20).
  for (size_t i = 1; i + 1 < ids.size(); ++i) EXPECT_EQ(ids[i], 20);
}

TEST(Tokenizer, NoTruncationMatchesHf) {
  Tokenizer t;
  std::string long_seq(1100, 'M');
  auto ids = t.Encode(long_seq, /*add_special=*/true, /*truncate=*/false);
  EXPECT_EQ(ids.size(), 1102u);
  EXPECT_EQ(ids.front(), 0);
  EXPECT_EQ(ids.back(), 2);
}

TEST(Tokenizer, DecodeRoundTripWithoutUnks) {
  Tokenizer t;
  std::string seq = "MKTGVAQERSILDPQNFY";
  auto ids = t.Encode(seq, /*add_special=*/false, /*truncate=*/true);
  EXPECT_EQ(t.Decode(ids), seq);
}

TEST(Tokenizer, DecodeSkipSpecials) {
  Tokenizer t;
  auto ids = t.Encode("MKTGV", /*add_special=*/true, /*truncate=*/true);
  EXPECT_EQ(t.Decode(ids, /*skip_special_tokens=*/false), "<cls>MKTGV<eos>");
  EXPECT_EQ(t.Decode(ids, /*skip_special_tokens=*/true), "MKTGV");
}

TEST(Tokenizer, TokenToIdAndIdToToken) {
  Tokenizer t;
  EXPECT_EQ(t.TokenToId("M"), 20);
  EXPECT_EQ(t.TokenToId("<mask>"), 32);
  EXPECT_EQ(t.TokenToId("nonsense"), Tokenizer::kUnkId);
  EXPECT_EQ(t.IdToToken(20), "M");
  EXPECT_EQ(t.IdToToken(999), "");
}
